text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,651 @@
+with the package itself (and not about it's listing in Construct Hub). The
+routing is based on metadata configured on the package itself, through the bug
+tracker URL, repository URL, and homepage URL.
+
+## Technical Design
+
+### Overview
+
+![Architecture Diagram](../images/cdk-construct-hub/architecture-diagram.png)
+
+### What is the high level implementation plan?
+
+#### Overview
+
+The Construct Hub is a relatively simple static web application. While the
+dataset it exposes (all Construct Libraries) may be relatively fast moving,
+there is no necessity for newly published packages (or versions thereof) to be
+indexed and presented on the application particularly quickly. A consistency
+window between a couple of hours and a day is probably acceptable.
+
+A simple data pipeline will ingest package publication events created by a
+source-specific adapter (e.g: by polling on the CouchDB instance replica for the
+[npmjs.com registry](https://skimdb.npmjs.com/registry)), and produce static
+artifacts the website is based off of. This pipeline is in particular
+responsible for extracting the `.jsii` assembly file from the npm packages, and
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. **Package Discovery:** The back-end for Construct Hub starts with a custom
+   built event source, which implements the logic necessary to detect new
+   packages that are relevant to the Construct Hub, and sends messages to an SQS
+   queue for further processing. Notifications have the following attributes:
+
+   Name          | Description
+   --------------|--------------------------------------------------------------
+   `assemblyUri` | The URL to an S3 Object containing the `.jsii` assembly included in the package
+   `time`        | The timestamp at which the version was created
+   `integrity`   | An integirity check for the complete record
+
+   The Construct Hub will have permissions to read from (and, if applicable, to
+   decrypt) the S3 bucket that is used to hold the object denoted by
+   `assemblyUri`. That bucket may have a Lifecycle configuration that cleans up
+   objects after a few days, as objects need not be retained there after they
+   have been successfully ingested by the Construct Hub (which will own and
+   maintain a copy of this artifact).
+
+   Construct Hub comes with a built-in integration function that listents to
+   changes on the [npmjs.com registry](https://skimdb.npmjs.com/registry)
+   database. Operators of private instances are able to interface with other
+   package registries (including private ones), such as CodeArtifact
+   repositories, etc.
+
+1. **Ingestion:** A Lambda function then picks up messages from the SQS queue
+   and prepares the artifacts consumed by the front-end application, stored in a
+   dedicated S3 bucket using the following key format:
+   `assemblies/${assembly.name}/v${assembly.version}/assembly.json`
```

Comment:
I am wondering if perhaps we should provision for additional annotations that the backend may want to add. For example, the classification which CDK a module works with might be some logic that runs during package processing and we would want to add that to the response to the client.

So maybe this file should look like this:

```json
{
  "name": ...,
  "version": ...,
  "manifest": { .jsii manifest },
  ... // additional fields that backend adds if needed
}
```

Basically what I am proposing is to store the .jsii manifest inside an "envelope" that will allow us to add additional fields by the backend without having to tweak the jsii manifest itself.

Reply:
Those annotations can be stored in S3 object metadata if they are small enough. These can be manipulated without ever touching the manifest itself. Now if we foresee we'd possibly need more metadata than fits in S3 object metadata... then yeah, envelope is a better solution.

Reply:
How will the client get the object metadata? Sounds like the wrong tool. All I am proposing is to return an object that contains the manifest as a field instead of the manifest directly, so we can later add additional fields to that object if we want. It's a common practice in protocol design...

Reply:
Object metadata is part of the headers you get with the object data.